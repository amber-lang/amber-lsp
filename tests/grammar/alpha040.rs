use std::fs::read_to_string;

use chumsky::error::Rich;
use insta::assert_debug_snapshot;

use amber_lsp::grammar::{
    alpha040::{lexer::Token, AmberCompiler, GlobalStatement, Spanned},
    LSPAnalysis, ParserResponse,
};

fn tokenize(input: &str) -> Vec<Spanned<Token>> {
    AmberCompiler::new().tokenize(input)
}

fn parse(
    tokens: &[Spanned<Token>],
) -> (Option<Vec<Spanned<GlobalStatement>>>, Vec<Rich<'_, String>>) {
    let ParserResponse {
        ast,
        errors,
        semantic_tokens: _,
    } = AmberCompiler::new().parse(tokens);

    let ast = match ast {
        amber_lsp::grammar::Grammar::Alpha040(ast) => ast,
        _ => panic!("Unexpected AST"),
    };

    (ast, errors)
}

fn parse_unwrap(tokens: &[Spanned<Token>]) -> Vec<Spanned<GlobalStatement>> {
    let (ast, errors) = parse(tokens);
    if !errors.is_empty() {
        panic!("Errors: {:?}", errors);
    }
    ast.unwrap()
}

#[test]
fn test_stdlib_array() {
    let stdlib = read_to_string("resources/alpha040/std/array.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_date() {
    let stdlib = read_to_string("resources/alpha040/std/date.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_env() {
    let stdlib = read_to_string("resources/alpha040/std/env.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_fs() {
    let stdlib = read_to_string("resources/alpha040/std/fs.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_http() {
    let stdlib = read_to_string("resources/alpha040/std/http.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_math() {
    let stdlib = read_to_string("resources/alpha040/std/math.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_text() {
    let stdlib = read_to_string("resources/alpha040/std/text.ab").unwrap();

    let tokens = tokenize(&stdlib);

    assert_debug_snapshot!(tokens);

    assert_debug_snapshot!(parse_unwrap(&tokens));
}

#[test]
fn test_unfinished_function_call() {
    let input = "
    import { array_contains } from \"std/array\"
    
    let x = [1, 2, 3]
    let y = 2
    
    echo array_contains(x, y)
    
    let line = 213
    
    for idx, line in lines(\"text.txt\") {
      echo line
    }
    
    // fun foo(x: Num, y: Text) {
    
    // }
    
    fun foo(x: Num, y: Text) {
    }
    
    // fun abc() {}
    
    array_contains([1, 2, 3],)
    ";

    let tokens = tokenize(input);
    assert_debug_snapshot!(tokens);

    let result = parse(&tokens);

    assert_debug_snapshot!(result);
}
