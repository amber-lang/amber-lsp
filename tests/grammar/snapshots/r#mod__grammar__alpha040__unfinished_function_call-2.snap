---
source: tests/grammar/alpha040.rs
expression: result
---
(
    Some(
        [
            (
                Import(
                    (
                        false,
                        5..11,
                    ),
                    (
                        "import",
                        5..11,
                    ),
                    (
                        ImportSpecific(
                            [
                                (
                                    "array_contains",
                                    14..28,
                                ),
                            ],
                        ),
                        12..30,
                    ),
                    (
                        "from",
                        31..35,
                    ),
                    (
                        "std/array",
                        36..47,
                    ),
                ),
                5..47,
            ),
            (
                Statement(
                    (
                        VariableInit(
                            (
                                "let",
                                57..60,
                            ),
                            (
                                "x",
                                61..62,
                            ),
                            (
                                Expression(
                                    (
                                        Array(
                                            [
                                                (
                                                    Number(
                                                        (
                                                            1.0,
                                                            66..67,
                                                        ),
                                                    ),
                                                    66..67,
                                                ),
                                                (
                                                    Number(
                                                        (
                                                            2.0,
                                                            69..70,
                                                        ),
                                                    ),
                                                    69..70,
                                                ),
                                                (
                                                    Number(
                                                        (
                                                            3.0,
                                                            72..73,
                                                        ),
                                                    ),
                                                    72..73,
                                                ),
                                            ],
                                        ),
                                        65..74,
                                    ),
                                ),
                                65..74,
                            ),
                        ),
                        57..74,
                    ),
                ),
                57..74,
            ),
            (
                Statement(
                    (
                        VariableInit(
                            (
                                "let",
                                79..82,
                            ),
                            (
                                "y",
                                83..84,
                            ),
                            (
                                Expression(
                                    (
                                        Number(
                                            (
                                                2.0,
                                                87..88,
                                            ),
                                        ),
                                        87..88,
                                    ),
                                ),
                                87..88,
                            ),
                        ),
                        79..88,
                    ),
                ),
                79..88,
            ),
            (
                Statement(
                    (
                        Echo(
                            (
                                "echo",
                                98..102,
                            ),
                            (
                                FunctionInvocation(
                                    [],
                                    (
                                        "array_contains",
                                        103..117,
                                    ),
                                    [
                                        (
                                            Var(
                                                (
                                                    "x",
                                                    118..119,
                                                ),
                                            ),
                                            118..119,
                                        ),
                                        (
                                            Var(
                                                (
                                                    "y",
                                                    121..122,
                                                ),
                                            ),
                                            121..122,
                                        ),
                                    ],
                                    None,
                                ),
                                103..123,
                            ),
                        ),
                        98..123,
                    ),
                ),
                98..123,
            ),
            (
                Statement(
                    (
                        VariableInit(
                            (
                                "let",
                                133..136,
                            ),
                            (
                                "line",
                                137..141,
                            ),
                            (
                                Expression(
                                    (
                                        Number(
                                            (
                                                213.0,
                                                144..147,
                                            ),
                                        ),
                                        144..147,
                                    ),
                                ),
                                144..147,
                            ),
                        ),
                        133..147,
                    ),
                ),
                133..147,
            ),
            (
                Statement(
                    (
                        IterLoop(
                            (
                                "for",
                                157..160,
                            ),
                            (
                                WithIndex(
                                    (
                                        "idx",
                                        161..164,
                                    ),
                                    (
                                        "line",
                                        166..170,
                                    ),
                                ),
                                161..170,
                            ),
                            (
                                "in",
                                171..173,
                            ),
                            (
                                FunctionInvocation(
                                    [],
                                    (
                                        "lines",
                                        174..179,
                                    ),
                                    [
                                        (
                                            Text(
                                                [
                                                    (
                                                        Text(
                                                            (
                                                                "text",
                                                                181..185,
                                                            ),
                                                        ),
                                                        181..185,
                                                    ),
                                                    (
                                                        Text(
                                                            (
                                                                ".",
                                                                185..186,
                                                            ),
                                                        ),
                                                        185..186,
                                                    ),
                                                    (
                                                        Text(
                                                            (
                                                                "txt",
                                                                186..189,
                                                            ),
                                                        ),
                                                        186..189,
                                                    ),
                                                ],
                                            ),
                                            180..190,
                                        ),
                                    ],
                                    None,
                                ),
                                174..191,
                            ),
                            (
                                Block(
                                    [],
                                    [
                                        (
                                            Echo(
                                                (
                                                    "echo",
                                                    200..204,
                                                ),
                                                (
                                                    Var(
                                                        (
                                                            "line",
                                                            205..209,
                                                        ),
                                                    ),
                                                    205..209,
                                                ),
                                            ),
                                            200..209,
                                        ),
                                    ],
                                ),
                                192..215,
                            ),
                        ),
                        157..215,
                    ),
                ),
                157..215,
            ),
            (
                Statement(
                    (
                        Comment(
                            "fun foo(x: Num, y: Text) {",
                        ),
                        225..255,
                    ),
                ),
                225..255,
            ),
            (
                Statement(
                    (
                        Comment(
                            "}",
                        ),
                        264..269,
                    ),
                ),
                264..269,
            ),
            (
                FunctionDefinition(
                    [],
                    (
                        false,
                        278..269,
                    ),
                    (
                        "fun",
                        278..281,
                    ),
                    (
                        "foo",
                        282..285,
                    ),
                    [
                        (
                            Typed(
                                (
                                    false,
                                    286..286,
                                ),
                                (
                                    "x",
                                    286..287,
                                ),
                                (
                                    Num,
                                    289..292,
                                ),
                            ),
                            286..292,
                        ),
                        (
                            Typed(
                                (
                                    false,
                                    294..293,
                                ),
                                (
                                    "y",
                                    294..295,
                                ),
                                (
                                    Text,
                                    297..301,
                                ),
                            ),
                            294..301,
                        ),
                    ],
                    None,
                    [],
                ),
                278..310,
            ),
            (
                Statement(
                    (
                        Comment(
                            "fun abc() {}",
                        ),
                        320..336,
                    ),
                ),
                320..336,
            ),
            (
                Statement(
                    (
                        Expression(
                            (
                                FunctionInvocation(
                                    [],
                                    (
                                        "array_contains",
                                        345..359,
                                    ),
                                    [
                                        (
                                            Array(
                                                [
                                                    (
                                                        Number(
                                                            (
                                                                1.0,
                                                                361..362,
                                                            ),
                                                        ),
                                                        361..362,
                                                    ),
                                                    (
                                                        Number(
                                                            (
                                                                2.0,
                                                                364..365,
                                                            ),
                                                        ),
                                                        364..365,
                                                    ),
                                                    (
                                                        Number(
                                                            (
                                                                3.0,
                                                                367..368,
                                                            ),
                                                        ),
                                                        367..368,
                                                    ),
                                                ],
                                            ),
                                            360..369,
                                        ),
                                    ],
                                    None,
                                ),
                                345..371,
                            ),
                        ),
                        345..371,
                    ),
                ),
                345..371,
            ),
        ],
    ),
    [],
)
